[Board] board application&infra 구현

vwx12345 opened the pull request:
### 💻 작업 내용
- Board의 application과 infrastructure 계층을 구현했습니다.

### ✨ 리뷰 포인트
-  PermissionCheckService와 BoardMapper 로 domain <-> entity 위주로 봐주시면 좋겠습니다.

### 📝 메모
- Query는 따로 분리해서 pr 올리겠습니다.
- BoardNamePrincipalImpl 에 NicknameConflictException 관련 에러가 spotless 적용으로 인해 잘못 올라갔습니다.

### 🎯 관련 이슈
closed #88

vwx12345 committed:
feat(member): BoardCommandService 정의 및 dto 정의

- BoardCommandService 정의
- BoardCreateCommand, BoardDeleteCommand, BoardCreateResult 정의

vwx12345 committed:
feat(member): PermissionCheckService 정의

- 권한 확인용 PermissionCheckService 정의

vwx12345 committed:
feat(member): PermissionCheckService 구현

- BoardCommandServiceImpl 구현
- BoardNotFoundException 정의

vwx12345 committed:
feat(member): BoardEntity 구현

vwx12345 committed:
feat(member): BoardMapper 구현

- Board domain 과 Board Entity 간 맵핑해주는 method 구현
- Board domain 에서 boardName 검증 없이 생성해주는 reconstruct 메서드 추가

vwx12345 committed:
feat(member): BoardEntityRepository 및 BoardRepositoryImpl 구현

vwx12345 committed:
feat(board): BoardNamePrincipalImpl 구현

vwx12345 committed:
feat(board): PermissionCheckService 구현

vwx12345 committed:
chore(Board): spotless 적용

vwx12345 committed:
docs(Board): 주석 작성

vwx12345 committed:
test(Board): BoardCommandService test 구현

vwx12345 committed:
test(Board): BoardMapper test 구현

codingbaraGo commented on src/main/java/com/simpleboard/board/board/infrastructure/PermissionCheckServiceImpl.java:
```diff
@@ -0,0 +1,38 @@
+            boardId,
+            memberId)
+        .retrieve()
+        .onStatus(HttpStatusCode::is4xxClientError, WebClientErrorHandler::handle4xx)
+        .onStatus(HttpStatusCode::is5xxServerError, WebClientErrorHandler::handle5xx)
```
Permission B.C에서 Exception을 반환하면 처리해주는 로직이 필요할 것 같습니다.
지금은 4XX 하나, 5XX 하나 총 2개의 에러로 처리해주는데, Permission B.C에서 반환한 에러의 종류마다 다르게 처리해주는 핸들링 로직이 필요할 것 같아요.
만약 유저가 권한이 없어 Permission B.C에서 권한없음 ServiceException을 반환했다면, Board B.C에서 이걸 처리해서 유저에게도 권한이 없다는 에러를 반환해야 할것 같네요.

vwx12345 replied:
권한 검증 실패는 Permission B.C에서 예외를 그대로 던지지 말고 ‘권한 없음’이라는 결과값으로 반환하는 게 맞다고 생각합니다.
Board B.C는 이 값을 받아 권한이 없다는 걸 확인한 뒤 이후 로직을 중단하거나 처리하면 됩니다. 만약 Board B.C에서 직접 예외를 잡아야 한다면, 굳이 internal API 방식으로 나눌 이유가 없어집니다.

codingbaraGo replied:
> 확인해보니 Permission쪽 internal API에서 boolean을 반환하는군요
> 그건 몰랐습니다.
> PermissionCheckServiceImpl에서 반환받은 값을 매핑하는 코드가 없어 exception을 던져 처리한다고 생각했습니다.
> 반환값이 있다면, 반환받은 값을 처리해주는 로직이 필요하지 않을까요?
> 지금 코드를 보면 요청을 보내기만 해서 반환값이 true이던 false이던 차이가 없어보여요.

확인해보니 Permission쪽 internal API에서 boolean을 반환하는군요
그건 몰랐습니다.
PermissionCheckServiceImpl에서 반환받은 값을 매핑하는 코드가 없어 exception을 던져 처리한다고 생각했습니다.
반환값이 있다면, 반환받은 값을 처리해주는 로직이 필요하지 않을까요?
지금 코드를 보면 요청을 보내기만 해서 반환값이 true이던 false이던 차이가 없어보여요.

vwx12345 replied:
> > 확인해보니 Permission쪽 internal API에서 boolean을 반환하는군요
> > 그건 몰랐습니다.
> > PermissionCheckServiceImpl에서 반환받은 값을 매핑하는 코드가 없어 exception을 던져 처리한다고 생각했습니다.
> > 반환값이 있다면, 반환받은 값을 처리해주는 로직이 필요하지 않을까요?
> > 지금 코드를 보면 요청을 보내기만 해서 반환값이 true이던 false이던 차이가 없어보여요.
> 
> 확인해보니 Permission쪽 internal API에서 boolean을 반환하는군요 그건 몰랐습니다. PermissionCheckServiceImpl에서 반환받은 값을 매핑하는 코드가 없어 exception을 던져 처리한다고 생각했습니다. 반환값이 있다면, 반환받은 값을 처리해주는 로직이 필요하지 않을까요? 지금 코드를 보면 요청을 보내기만 해서 반환값이 true이던 false이던 차이가 없어보여요.

제가 확실히 반환값을 처리를 안해서 internal API와 관련한 오해하셨던 것 같네요. 현창님 말대로 반환값 처리하는 방식으로 변경해서 다시 pr하겠습니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/infrastructure/BoardNamePrincipalImpl.java:
```diff
@@ -0,0 +1,19 @@
+package com.simpleboard.board.board.infrastructure;
+
+import com.simpleboard.board.board.domain.board.repository.BoardRepository;
+import com.simpleboard.board.board.domain.board.service.BoardNamePrincipal;
+import com.simpleboard.board.board.domain.board.vo.BoardName;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class BoardNamePrincipalImpl implements BoardNamePrincipal {
```
BoardNamePrincipal은 도메인 로직으로 알고있는데 infra 계층에 위치해있네요.
클래스 내부의 코드를 봐도 infra 계층에 있을 이유는 없어보이는데 특별한 이유가 있을까요?

vwx12345 replied:
> BoardNamePrincipal은 도메인 로직으로 알고있는데 infra 계층에 위치해있네요. 클래스 내부의 코드를 봐도 infra 계층에 있을 이유는 없어보이는데 특별한 이유가 있을까요?

앞으로 캐시나 다른 저장소를 활용한 구현으로도 교체 가능하다는 점을 고려해, 인터페이스는 도메인에 두고 구현체는 인프라에 두는 구조를 선택했습니다.

다만 지적해주신 부분처럼 단순히 BoardRepository 인터페이스에만 의존할지, 아니면 인프라 구현체(BoardRepositoryImpl)를 직접 의존하는 것이 더 적절한지 고민해볼 필요가 있습니다.

SnoopyComp replied:
> > BoardNamePrincipal은 도메인 로직으로 알고있는데 infra 계층에 위치해있네요. 클래스 내부의 코드를 봐도 infra 계층에 있을 이유는 없어보이는데 특별한 이유가 있을까요?
> 
> 앞으로 캐시나 다른 저장소를 활용한 구현으로도 교체 가능하다는 점을 고려해, 인터페이스는 도메인에 두고 구현체는 인프라에 두는 구조를 선택했습니다.
> 
> 다만 지적해주신 부분처럼 단순히 BoardRepository 인터페이스에만 의존할지, 아니면 인프라 구현체(BoardRepositoryImpl)를 직접 의존하는 것이 더 적절한지 고민해볼 필요가 있습니다.

저는 BoardNamePrincipal 이 domain principal 이라고 생각했는데 아닌가요?
BoardRepository는 command 레포지토리였던걸로 아는데, 구현체에 의존한다 해도 특별히 달라지는건 없을 것 같아요.
도메인 규칙의 경우 도메인 서비스 클래스처럼 Domain 영역에 위치해야 합니다.

도메인 규칙 내부에서 repository를 사용한다 하더라도, DI를 통해 도메인 영역에 존재하는 repository를 주입받아 사용해야지, repository를 쓰기 위해 도메인 규칙 클래스가 infra 영역으로 가서는 안된다고 생각해요.

캐시 도입이나 저장소의 교체가 걱정된다면 
- BoardNamePrincipal 전용 도메인 레포지토리(인터페이스) 사용
- 기존의 도메인(혹은 Query용) 레포지토리에 전용 메서드 추가
와 같은 방법을 고려해보시는건 어떨까요?

codingbaraGo replied:
> > BoardNamePrincipal은 도메인 로직으로 알고있는데 infra 계층에 위치해있네요. 클래스 내부의 코드를 봐도 infra 계층에 있을 이유는 없어보이는데 특별한 이유가 있을까요?
> 
> 앞으로 캐시나 다른 저장소를 활용한 구현으로도 교체 가능하다는 점을 고려해, 인터페이스는 도메인에 두고 구현체는 인프라에 두는 구조를 선택했습니다.
> 
> 다만 지적해주신 부분처럼 단순히 BoardRepository 인터페이스에만 의존할지, 아니면 인프라 구현체(BoardRepositoryImpl)를 직접 의존하는 것이 더 적절한지 고민해볼 필요가 있습니다.

저는 BoardNamePrincipal 이 domain principal 이라고 생각했는데 아닌가요?
BoardRepository는 command 레포지토리였던걸로 아는데, 구현체에 의존한다 해도 특별히 달라지는건 없을 것 같아요.
도메인 규칙의 경우 도메인 서비스 클래스처럼 Domain 영역에 위치해야 합니다.

도메인 규칙 내부에서 repository를 사용한다 하더라도, DI를 통해 도메인 영역에 존재하는 repository를 주입받아 사용해야지, repository를 쓰기 위해 도메인 규칙 클래스가 infra 영역으로 가서는 안된다고 생각해요.

캐시 도입이나 저장소의 교체가 걱정된다면

BoardNamePrincipal 전용 도메인 레포지토리(인터페이스) 사용
기존의 도메인(혹은 Query용) 레포지토리에 전용 메서드 추가
와 같은 방법을 고려해보시는건 어떨까요?

vwx12345 replied:
저는 BoardNamePrincipal은 Domain Principal로서 단순히 isUnique라는 규칙만 정의하면 충분하다고 생각합니다. 그 규칙을 어떻게 구현할지는 Infra 계층의 역할입니다. 실제로 Domain Repository에서도 “해야 할 동작”만 정의하고, 실구현체는 JPA든 MyBatis든 Infra 계층에서 맡기는 것과 같은 맥락입니다.

현재 구조는 캐시를 사용하든, 레포지토리를 사용하든 isUnique를 구현해주면 됩니다. 그런데 만약 도메인 코드에서 캐시와 레포지토리를 동시에 고려해야 한다면, 결국 도메인 코드 자체를 수정해야 하는 문제가 생깁니다.

따라서 도메인 규칙은 Domain 계층에 두고, 구현은 Infra 계층에서 담당하는 지금의 방식이 더 낫다고 생각합니다.

codingbaraGo replied:
> 저는 BoardNamePrincipal은 Domain Principal로서 단순히 isUnique라는 규칙만 정의하면 충분하다고 생각합니다. 그 규칙을 어떻게 구현할지는 Infra 계층의 역할입니다. 실제로 Domain Repository에서도 “해야 할 동작”만 정의하고, 실구현체는 JPA든 MyBatis든 Infra 계층에서 맡기는 것과 같은 맥락입니다.
> 
> 현재 구조는 캐시를 사용하든, 레포지토리를 사용하든 isUnique를 구현해주면 됩니다. 그런데 만약 도메인 코드에서 캐시와 레포지토리를 동시에 고려해야 한다면, 결국 도메인 코드 자체를 수정해야 하는 문제가 생깁니다.
> 
> 따라서 도메인 규칙은 Domain 계층에 두고, 구현은 Infra 계층에서 담당하는 지금의 방식이 더 낫다고 생각합니다.

Repository 조회를 통해 중복되는 board name이 있는지 확인하는 로직은 비지니스 로직입니다.
캐시와 DB 라이브러리의 종류는 repository에서 처리해야 하는 부분입니다.
도메인 규칙 클래스는 관련 내용을 몰라야 하고, 관련 책임은 전부 repository에서 가져야 합니다.

도메인 서비스 클래스를 생각하면 좋을것같네요.
도메인 서비스 클래스의 경우에도 레포지토리를 호출하여 비지니스 로직을 수행하지만, 레포지토리와 엄격하게 분리되어 레포지토리 내부의 업무를 알 수 없도록 설계하죠.
도메인 서비스 클래스의 구현체도 도메인 계층에 구현하고요.
도메인 규칙 또한 비슷하게 생각해주시면 될 것 같습니다.

비지니스 로직인 Domain principal을 인프라 계층에 내리는 것은 계층 분리가 명료하지 못한게 아닌가 합니다.

추가로, 캐싱의 경우 다양한 테이블, 혹은 컬럼들에 대해 이루어질 텐데 그때마다 관련 도메인 서비스, 혹은 응용 서비스를 인프라 계층으로 내릴수는 없습니다.
캐싱과 DB라이브러리 관련 책임은 전부 레포지토리에서 가져가는것이 이상적인 계층분리의 형태가 아닐까 합니다

vwx12345 replied:
> > 저는 BoardNamePrincipal은 Domain Principal로서 단순히 isUnique라는 규칙만 정의하면 충분하다고 생각합니다. 그 규칙을 어떻게 구현할지는 Infra 계층의 역할입니다. 실제로 Domain Repository에서도 “해야 할 동작”만 정의하고, 실구현체는 JPA든 MyBatis든 Infra 계층에서 맡기는 것과 같은 맥락입니다.
> > 현재 구조는 캐시를 사용하든, 레포지토리를 사용하든 isUnique를 구현해주면 됩니다. 그런데 만약 도메인 코드에서 캐시와 레포지토리를 동시에 고려해야 한다면, 결국 도메인 코드 자체를 수정해야 하는 문제가 생깁니다.
> > 따라서 도메인 규칙은 Domain 계층에 두고, 구현은 Infra 계층에서 담당하는 지금의 방식이 더 낫다고 생각합니다.
> 
> Repository 조회를 통해 중복되는 board name이 있는지 확인하는 로직은 비지니스 로직입니다. 캐시와 DB 라이브러리의 종류는 repository에서 처리해야 하는 부분입니다. 도메인 규칙 클래스는 관련 내용을 몰라야 하고, 관련 책임은 전부 repository에서 가져야 합니다.
> 
> 도메인 서비스 클래스를 생각하면 좋을것같네요. 도메인 서비스 클래스의 경우에도 레포지토리를 호출하여 비지니스 로직을 수행하지만, 레포지토리와 엄격하게 분리되어 레포지토리 내부의 업무를 알 수 없도록 설계하죠. 도메인 서비스 클래스의 구현체도 도메인 계층에 구현하고요. 도메인 규칙 또한 비슷하게 생각해주시면 될 것 같습니다.
> 
> 비지니스 로직인 Domain principal을 인프라 계층에 내리는 것은 계층 분리가 명료하지 못한게 아닌가 합니다.
> 
> 추가로, 캐싱의 경우 다양한 테이블, 혹은 컬럼들에 대해 이루어질 텐데 그때마다 관련 도메인 서비스, 혹은 응용 서비스를 인프라 계층으로 내릴수는 없습니다. 캐싱과 DB라이브러리 관련 책임은 전부 레포지토리에서 가져가는것이 이상적인 계층분리의 형태가 아닐까 합니다

현창님 말이 맞는 것 같습니다.
[Member] Member Infra 계층 구현 #120 에 의견 남겨놨습니다.

codingbaraGo replied:
> > > 저는 BoardNamePrincipal은 Domain Principal로서 단순히 isUnique라는 규칙만 정의하면 충분하다고 생각합니다. 그 규칙을 어떻게 구현할지는 Infra 계층의 역할입니다. 실제로 Domain Repository에서도 “해야 할 동작”만 정의하고, 실구현체는 JPA든 MyBatis든 Infra 계층에서 맡기는 것과 같은 맥락입니다.
> > > 현재 구조는 캐시를 사용하든, 레포지토리를 사용하든 isUnique를 구현해주면 됩니다. 그런데 만약 도메인 코드에서 캐시와 레포지토리를 동시에 고려해야 한다면, 결국 도메인 코드 자체를 수정해야 하는 문제가 생깁니다.
> > > 따라서 도메인 규칙은 Domain 계층에 두고, 구현은 Infra 계층에서 담당하는 지금의 방식이 더 낫다고 생각합니다.
> > 
> > 
> > Repository 조회를 통해 중복되는 board name이 있는지 확인하는 로직은 비지니스 로직입니다. 캐시와 DB 라이브러리의 종류는 repository에서 처리해야 하는 부분입니다. 도메인 규칙 클래스는 관련 내용을 몰라야 하고, 관련 책임은 전부 repository에서 가져야 합니다.
> > 도메인 서비스 클래스를 생각하면 좋을것같네요. 도메인 서비스 클래스의 경우에도 레포지토리를 호출하여 비지니스 로직을 수행하지만, 레포지토리와 엄격하게 분리되어 레포지토리 내부의 업무를 알 수 없도록 설계하죠. 도메인 서비스 클래스의 구현체도 도메인 계층에 구현하고요. 도메인 규칙 또한 비슷하게 생각해주시면 될 것 같습니다.
> > 비지니스 로직인 Domain principal을 인프라 계층에 내리는 것은 계층 분리가 명료하지 못한게 아닌가 합니다.
> > 추가로, 캐싱의 경우 다양한 테이블, 혹은 컬럼들에 대해 이루어질 텐데 그때마다 관련 도메인 서비스, 혹은 응용 서비스를 인프라 계층으로 내릴수는 없습니다. 캐싱과 DB라이브러리 관련 책임은 전부 레포지토리에서 가져가는것이 이상적인 계층분리의 형태가 아닐까 합니다
> 
> 현창님 말이 맞는 것 같습니다. [Member] Member Infra 계층 구현 #120 에 의견 남겨놨습니다.

이건 그냥 팁인데, 코멘트 우측상단의 ... 에서 "Copy link"를 클릭하시면 특정 코멘트에 대한 링크를 사용할 수 있습니다!
https://github.com/SimpleBoardTeam/simpleBoard-server/pull/120#discussion_r2321527943

codingbaraGo commented on src/main/java/com/simpleboard/board/board/infrastructure/jpa/mapper/BoardMapper.java:
```diff
@@ -0,0 +1,33 @@
+
+  /** Domain -> JPA Entity */
+  public BoardEntity toEntity(Board domain) {
+    if (domain == null) return null;
```
Exception을 던지는 편이 안전하지 않을까 합니다

vwx12345 replied:
> Exception을 던지는 편이 안전하지 않을까 합니다

다른 로직을 확인해보니 도메인 검증 로직은 별도로 두지 않고 있어서, 통일성을 위해 해당 부분은 제거하도록 하겠습니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/infrastructure/jpa/repository/BoardRepositoryImpl.java:
```diff
@@ -0,0 +1,49 @@
+        .map(BoardEntity::getBoardName)
+        .map(BoardName::of)
+        .map(BoardName::normalized)
+        .anyMatch(normalizedName::equals);
```
전체조회는 상당히 위험한 쿼리입니다.
dc에는 65000개 이상의 게시판(갤러리)들이 존재하고, reddit은 21년도 기준 100만개가 넘어갑니다.
이런 대용량 데이터를 전체조회하면 DB와 램 메모리 양쪽 모두에 큰 부하가 옵니다.
또, stream으로 전체탐색을 진행하니 cpu도 과하게 사용하게 되고, 무엇보다 응답속도가 현저하게 떨어질 것입니다.
다른 방법을 사용해야 할 것 같습니다.

vwx12345 replied:
> 전체조회는 상당히 위험한 쿼리입니다. dc에는 65000개 이상의 게시판(갤러리)들이 존재하고, reddit은 21년도 기준 100만개가 넘어갑니다. 이런 대용량 데이터를 전체조회하면 DB와 램 메모리 양쪽 모두에 큰 부하가 옵니다. 또, stream으로 전체탐색을 진행하니 cpu도 과하게 사용하게 되고, 무엇보다 응답속도가 현저하게 떨어질 것입니다. 다른 방법을 사용해야 할 것 같습니다.

현재로서는 boardName 기준 인덱스를 활용하는 것 외에는 마땅한 대안이 떠오르지 않아, 게시판 수가 많지 않은 현 단계에서는 이대로 두고 v2에서 개선하는 방향으로 고려하겠습니다. 혹시 더 좋은 아이디어가 있다면 공유 부탁드립니다.

codingbaraGo replied:
> > 전체조회는 상당히 위험한 쿼리입니다. dc에는 65000개 이상의 게시판(갤러리)들이 존재하고, reddit은 21년도 기준 100만개가 넘어갑니다. 이런 대용량 데이터를 전체조회하면 DB와 램 메모리 양쪽 모두에 큰 부하가 옵니다. 또, stream으로 전체탐색을 진행하니 cpu도 과하게 사용하게 되고, 무엇보다 응답속도가 현저하게 떨어질 것입니다. 다른 방법을 사용해야 할 것 같습니다.
> 
> 현재로서는 boardName 기준 인덱스를 활용하는 것 외에는 마땅한 대안이 떠오르지 않아, 게시판 수가 많지 않은 현 단계에서는 이대로 두고 v2에서 개선하는 방향으로 고려하겠습니다. 혹시 더 좋은 아이디어가 있다면 공유 부탁드립니다.

일단 이런 대규모 조회는 Query 레포지토리를 사용하는 편이 좋지 않을까 합니다.
또, 전체조회 쿼리에서는 인덱스 또한 작동하지 않기 때문에 인덱스를 통한 최적화도 불가능합니다.

사실 board가 1000개정도만 되어도 살짝 불안해지는 코드입니다.

정규화를 거친 문자열 저장을 원하신다면 정규화 문자열과 raw 문자열 두개 모두 저장을 하거나, sql 함수를 통해 검색 쿼리가 정규화까지 고려하여 조회하도록 하여야하지 않을까 합니다.
jpql로 가능한지는 잘 모르겠네요.

지금처럼 전부 메모리로 불러서 stream을 통해 선형 탐색을 하는 방식은 정말 지양해야하는 방법이지 않나 합니다.

codingbaraGo replied:
> > > 전체조회는 상당히 위험한 쿼리입니다. dc에는 65000개 이상의 게시판(갤러리)들이 존재하고, reddit은 21년도 기준 100만개가 넘어갑니다. 이런 대용량 데이터를 전체조회하면 DB와 램 메모리 양쪽 모두에 큰 부하가 옵니다. 또, stream으로 전체탐색을 진행하니 cpu도 과하게 사용하게 되고, 무엇보다 응답속도가 현저하게 떨어질 것입니다. 다른 방법을 사용해야 할 것 같습니다.
> > 
> > 
> > 현재로서는 boardName 기준 인덱스를 활용하는 것 외에는 마땅한 대안이 떠오르지 않아, 게시판 수가 많지 않은 현 단계에서는 이대로 두고 v2에서 개선하는 방향으로 고려하겠습니다. 혹시 더 좋은 아이디어가 있다면 공유 부탁드립니다.
> 
> 일단 이런 대규모 조회는 Query 레포지토리를 사용하는 편이 좋지 않을까 합니다. 또, 전체조회 쿼리에서는 인덱스 또한 작동하지 않기 때문에 인덱스를 통한 최적화도 불가능합니다.
> 
> 사실 board가 1000개정도만 되어도 살짝 불안해지는 코드입니다.
> 
> 정규화를 거친 문자열 저장을 원하신다면 정규화 문자열과 raw 문자열 두개 모두 저장을 하거나, sql 함수를 통해 검색 쿼리가 정규화까지 고려하여 조회하도록 하여야하지 않을까 합니다. jpql로 가능한지는 잘 모르겠네요.
> 
> 지금처럼 전부 메모리로 불러서 stream을 통해 선형 탐색을 하는 방식은 정말 지양해야하는 방법이지 않나 합니다.

혹시 작업량 이슈로 pending해 두고 싶으시다면 todo만 붙여두고 나중에 리펙토링 해도 괜찮을것 같긴 하네요

vwx12345 replied:
> > > > 전체조회는 상당히 위험한 쿼리입니다. dc에는 65000개 이상의 게시판(갤러리)들이 존재하고, reddit은 21년도 기준 100만개가 넘어갑니다. 이런 대용량 데이터를 전체조회하면 DB와 램 메모리 양쪽 모두에 큰 부하가 옵니다. 또, stream으로 전체탐색을 진행하니 cpu도 과하게 사용하게 되고, 무엇보다 응답속도가 현저하게 떨어질 것입니다. 다른 방법을 사용해야 할 것 같습니다.
> > > 
> > > 
> > > 현재로서는 boardName 기준 인덱스를 활용하는 것 외에는 마땅한 대안이 떠오르지 않아, 게시판 수가 많지 않은 현 단계에서는 이대로 두고 v2에서 개선하는 방향으로 고려하겠습니다. 혹시 더 좋은 아이디어가 있다면 공유 부탁드립니다.
> > 
> > 
> > 일단 이런 대규모 조회는 Query 레포지토리를 사용하는 편이 좋지 않을까 합니다. 또, 전체조회 쿼리에서는 인덱스 또한 작동하지 않기 때문에 인덱스를 통한 최적화도 불가능합니다.
> > 사실 board가 1000개정도만 되어도 살짝 불안해지는 코드입니다.
> > 정규화를 거친 문자열 저장을 원하신다면 정규화 문자열과 raw 문자열 두개 모두 저장을 하거나, sql 함수를 통해 검색 쿼리가 정규화까지 고려하여 조회하도록 하여야하지 않을까 합니다. jpql로 가능한지는 잘 모르겠네요.
> > 지금처럼 전부 메모리로 불러서 stream을 통해 선형 탐색을 하는 방식은 정말 지양해야하는 방법이지 않나 합니다.
> 
> 혹시 작업량 이슈로 pending해 두고 싶으시다면 todo만 붙여두고 나중에 리펙토링 해도 괜찮을것 같긴 하네요

Normalized 값도 저장해서 existsByNormalizedName 를 통해 확인하는 방식으로 변경했습니다.

codingbaraGo reviewed (commented):
고생하셨습니다.

vwx12345 committed:
refactor(board): 검증 로직 통일을 위해 return 처리 제거

codingbaraGo reviewed (commented):
null 체크는 Mapper 밖에서 해준다고 생각하면 될까요?

vwx12345 committed:
refactor(board): BoardNamePrincipal 구조 변경

- 도메인 계층에 BoardNamePrincipal 를 인터페이스로 정의하고  인프라 계층에서 구현하는 방식 대신 도메인 계층에서 BoardRepository 를 통해 구현하는 방식이 DDD 관점에 맞다고 판단하여 변경했습니다.

vwx12345 committed:
fix(board): PermissionCheckServiceImpl 수정

- checkBoardDeletePermission 에서 반환 값을 처리하는 방식으로 수정했습니다.

vwx12345 committed:
chore(board): spotless apply

vwx12345 committed:
Merge branch 'develop' into feat/board-board-application&infra/#88

vwx12345 committed:
refactor(board): BoardName 정규화

- BoardName의 정규화 값 또한 DB에 저장하는 식으로 변경했습니다.
- 정규화된 BoardName을 통해 중복 체크 로직을 개선했습니다.

vwx12345 committed:
fix(board): BoardMapperTest 수정

- toDomain_maps_all_fields 에서 String 을 BoardName 으로 변경하고 테스트에서 String 으로 테스트 하는 문제를 수정했습니다.

vwx12345 committed:
chore(board): spotless 적용

vwx12345 committed:
fix(board): BoardMapper null 처리 추가

vwx12345 commented:
> null 체크는 Mapper 밖에서 해준다고 생각하면 될까요?

레퍼지토리에서 Optional 로 처리하기 때문에 null 이면 다시 null 을 return 하는 방식으로 코드를 변경했습니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/infrastructure/webclient/PermissionCheckServiceImpl.java:
```diff
@@ -23,16 +23,15 @@ public void checkBoardCreatePermission(Long memberId) {
+        .bodyToMono(CanDeleteResponse.class)
         .block();
+
+    if (response == null || !response.canDelete()) {
+      throw new BoardDeleteNotAllowedException();
+    }
```
response 값이 null이 돌아올 경우, internal api를 이용한 통신 과정에서 에러가 발생했기 때문일것입니다.
원래 api 명세상으로는 true or false를 반환해야하니까요.
이럴 때 다시 요청을 발송하거나, 로그를 남기거나, 사용자에게 에러 메시지를 반환하는 로직도 필요하겠군요...

다만, board 삭제 권한 체크의 경우, 
권한이 있는 사용자엑에 사용자에게 삭제 권한이 없다는 에러를 반환하면 사용자가 "왜 안되지?" 하고 삭제 요청을 다시 보낼테니 지금 상태로도 크리티컬한 이슈가 발생할것같진 않습니다.

이 부분은 고도화 기간에 수정하면 좋을 것 같아요.

vwx12345 replied:
> response 값이 null이 돌아올 경우, internal api를 이용한 통신 과정에서 에러가 발생했기 때문일것입니다. 원래 api 명세상으로는 true or false를 반환해야하니까요. 이럴 때 다시 요청을 발송하거나, 로그를 남기거나, 사용자에게 에러 메시지를 반환하는 로직도 필요하겠군요...
> 
> 다만, board 삭제 권한 체크의 경우, 권한이 있는 사용자엑에 사용자에게 삭제 권한이 없다는 에러를 반환하면 사용자가 "왜 안되지?" 하고 삭제 요청을 다시 보낼테니 지금 상태로도 크리티컬한 이슈가 발생할것같진 않습니다.
> 
> 이 부분은 고도화 기간에 수정하면 좋을 것 같아요.

사실 400 에러와 500 에러는 따로 처리해줘서 과연 null 값이 들어올 일이 있을까 싶지만 혹시나 해서 추가한 로직이었습니다. 고도화 기간 때 좀 더 생각해볼 필요는 있겠네요.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/infrastructure/webclient/dto/CanDeleteResponse.java:
```diff
@@ -0,0 +1,4 @@
+package com.simpleboard.board.board.infrastructure.webclient.dto;
+
+public record CanDeleteResponse(boolean canDelete) {}
+
```
그나저나 DTO의 종류는 끊임없이 늘어나는군요...
네이밍이 정말 제일 힘드네요

vwx12345 replied:
> 그나저나 DTO의 종류는 끊임없이 늘어나는군요... 네이밍이 정말 제일 힘드네요

DTO 부터 에러명, 함수명 등등 정말 쉽지 않은 것 같아요.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/domain/board/service/BoardNamePrincipal.java:
```diff
@@ -23,8 +28,8 @@ public interface BoardNamePrincipal {
    */
-  default void ensureUnique(BoardName name) {
-    if (!isUnique(name)) {
+  public void ensureUnique(String name) {
+    if (boardRepository.existsByNameNormalized(name)) {
```
argument로 들어오는 변수의 네이밍을 nomarlizedName으로 바꾸어준다면 
"정규화 처리 된 board name만 인자로 주어야 한다" 라는 규칙이 좀 더 명확해지지 않을까 하는 생각이 들어요.

vwx12345 replied:
> argument로 들어오는 변수의 네이밍을 nomarlizedName으로 바꾸어준다면 "정규화 처리 된 board name만 인자로 주어야 한다" 라는 규칙이 좀 더 명확해지지 않을까 하는 생각이 들어요.

좋은 피드백 감사합니다. 해당 부분 반영해서 다시 올리겠습니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/infrastructure/jpa/entity/BoardEntity.java:
```diff
@@ -30,6 +31,9 @@ public BoardEntity(Long boardId, String boardName, Long managerId) {
   @Column(name = "board_name", nullable = false, unique = true, length = 100)
   private String boardName;
 
+  @Column(name = "normalized_name", nullable = false, unique = true, length = 100)
+  private String normalizedName;
+
```
굿입니다. 이 편이 여러모로 좋을 것 같네요.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/infrastructure/jpa/repository/BoardRepositoryImpl.java:
```diff
@@ -35,11 +35,7 @@ public Optional<Board> findById(Long boardId) {
-        .map(BoardName::normalized)
-        .anyMatch(normalizedName::equals);
+    return boardEntityRepository.existsByNormalizedName(normalizedName);
   }
```
굿입니다.

codingbaraGo commented on src/test/java/com/simpleboard/board/board/infrastructure/jpa/mapper/BoardMapperTest.java:
```diff
@@ -50,10 +50,10 @@ void toEntity_maps_all_fields() {
     Long managerId = 2002L;
 
-    BoardEntity entity = new BoardEntity(boardId, boardName, managerId);
+    BoardEntity entity = new BoardEntity(boardId, boardName.toString(), boardName.normalized(), managerId);
```
이제 toString()과 normalized()가 확실하게 분리된 역할을 가지게 되었네요.
사실 처음에 toString()이라는 메서드 명이 조금 애매하지 않나 하는 생각도 들었는데 normalized랑 같이 보니 깔끔한 것 같습니다.

codingbaraGo reviewed (approved):
고생하셨습니다!

vwx12345 committed:
refactor(board): BoardNamePrincipal 내부 변수명 변경

- 규칙을 분명하기 위해 "name" 에서 "normalizedName" 으로 변수명 변경

codingbaraGo reviewed (approved):
👍
